### PR TITLE
Octavia umbrella controller

### PR DESCRIPTION
**Pull request by beagles**
When implementing the controllers for the non-API amphora provider related services, it became apparent that:

- functionality that was relevant to all controllers was encapsulated into the API controller (db-sync)
- from the perspective of install_yamls and the openstack meta operator, the precedent is for there to be a single umbrella controller that distributed reconcile operations among specialized controllers (see cinder and nova)

To bring Octavia in line we need to implement an "octavia" controller that delegates to the controllers for the api, worker, etc.. This PR is the "refactor" part, after which the WIP for the worker, housekeeping, healthmanager, will be rebased on

### CODE REVIEW

File: templates/octavia/config/octavia.conf [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+bind_port=9876
+auth_strategy=keystone
+# enabled_provider_drivers=amphora: The Octavia Amphora driver.,octavia: Deprecated alias of the Octavia Amphora driver.
+healthcheck_enabled=True
+tls_cipher_prohibit_list=
+default_listener_tls_versions=TLSv1.2,TLSv1.3
+default_pool_tls_versions=TLSv1.2,TLSv1.3
+[database]
+[health_manager]
+health_update_threads=4
```
Comment (johnsom) [inline templates/octavia/config/octavia.conf:19]: This seems very low, are you sure this is the right value?
This is an important value for the scalability of the number of amphora instances the deployment can handle.

File: templates/octavia/config/octavia.conf [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+# password=FIXMEpw3
+project_name=service
+project_domain_name=Default
+user_domain_name=Default
+auth_type=password
+# memcache_use_advanced_pool=True
+# memcached_servers=FIXMEhost1:11211
+# region_name=regionOne
+# interface=internal
+[certificates]
```
Comment (johnsom) [inline templates/octavia/config/octavia.conf:35]: We are moving to the dual CA model here right?
That is long standing tech debt in tripleo

**Review by johnsom — COMMENTED** [review]
I didn't go through all of this or all of the configuration settings that maybe should be here, but noted a few things.

File: templates/octavia/config/octavia.conf (hunk 1 above)
Comment (beagles) [reply templates/octavia/config/octavia.conf:19]: Good point. I'll revise all of the configs when adding those services

File: templates/octavia/config/octavia.conf (hunk 2 above)
Comment (beagles) [reply templates/octavia/config/octavia.conf:35]: Yes.

File: config/samples/kustomization.yaml [hunk 3]
```diff
@@ -1,4 +1,5 @@
 ## Append samples you want in your CSV to this file as resources ##
 resources:
 - octavia_v1beta1_octaviaapi.yaml
```
Comment (weinimo) [inline config/samples/kustomization.yaml:3]: Do we still need octavia_v1beta1_octaviaapi.yaml as a separate yaml file or should be better merge it into one?

**Comment by weinimo** [comment]
When I create a new deployment with this PR, the octavia-api pod gets stuck in Init:CrashLoopBackOff with the error

> + export PASSWORD=12345678
> + PASSWORD=12345678
> /usr/local/bin/container-scripts/init.sh: line 23: DatabaseHost: Please specify a DatabaseHost variable.

I haven't figured out what broke it though.

**Comment by weinimo** [comment]
The octavia operator itself fails too with:

```
2023-02-23T10:47:28.976Z        INFO    controllers.Octavia     Reconciling Service init
2023-02-23T10:47:28.981Z        ERROR   Reconciler error        {"controller": "octavia", "controllerGroup": "octavia.openstack.org", "controllerKind": "Octavia", "Octavia": {"name":"octavia","namespace":"openstack"}, "namespace": "openstack", "name": "octavia", "reconcileID": "c7ae6e34-42c9-411d-b19e-cc2d6685962b", "error": "Error create or update DB object octavia *v1beta1.MariaDBDatabase openstack/octavia: Object openstack/octavia is already owned by another OctaviaAPI controller octavia"}sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler        /home/stack/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:326sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem        /home/stack/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:273
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2        /home/stack/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.13.1/pkg/internal/controller/controller.go:234
```

File: pkg/octaviaapi/deployment.go [hunk 4]
```diff
@@ -137,25 +142,25 @@ func Deployment(
...
 	)
 	if instance.Spec.NodeSelector != nil && len(instance.Spec.NodeSelector) > 0 {
 		deployment.Spec.Template.Spec.NodeSelector = instance.Spec.NodeSelector
 	}
 
-	initContainerDetails := APIDetails{
+	initContainerDetails := octavia.APIDetails{
 		ContainerImage:       instance.Spec.ContainerImage,
-		DatabaseHost:         instance.Status.DatabaseHostname,
+		DatabaseHost:         instance.Spec.DatabaseHostname,
```
Comment (weinimo) [inline pkg/octaviaapi/deployment.go:155]:
```suggestion
		DatabaseHost:         instance.Status.DatabaseHostname,
```

The change in this line seems to have caused the octavia-api pod to get stuck in Init:CrashLoopBackOff with the error

```
+ export PASSWORD=12345678
+ PASSWORD=12345678
/usr/local/bin/container-scripts/init.sh: line 23: DatabaseHost: Please specify a DatabaseHost variable.
```

**Review by weinimo — CHANGES_REQUESTED** [review]
Still doing some tests

File: config/samples/kustomization.yaml (hunk 3 above)
Comment (beagles) [reply config/samples/kustomization.yaml:3]: Good point. Let's do away with it.

File: pkg/octaviaapi/deployment.go (hunk 4 above)
Comment (beagles) [reply pkg/octaviaapi/deployment.go:155]: I'll investigate this. AFAICT this bit of code is not different than what I've been using and has been working fine.

**Comment by beagles** [comment]
Note that this also requires an change to install_yamls. See https://github.com/openstack-k8s-operators/install_yamls/pull/119

File: pkg/octaviaapi/deployment.go (hunk 4 above)
Comment (beagles) [reply pkg/octaviaapi/deployment.go:155]: I think it's likely that an outdated install_yamls was used. The database hostname is passed to OctaviaAPI by the Octavia instance through it's spec, not it's status.

File: controllers/octavia_controller.go [hunk 5]
```diff
@@ -0,0 +1,594 @@
...
+	r.Log.Info("Reconciling Service upgrade")
+
+	// TODO: should have major version upgrade tasks
+	// -delete dbsync hash from status to rerun it?
+
+	r.Log.Info("Reconciled Service upgrade successfully")
+	return ctrl.Result{}, nil
+}
+
+// TODO(tweining): implement
```
Comment (weinimo) [inline controllers/octavia_controller.go:342]: I think we could remove this comment. The basic functionality is there.

**Review by weinimo — APPROVED** [review]
/lgtm
Thanks!

**Pull request merged**